src/backend/distributed/planner/multi_join_order.c

```diff
@@ -1276,7 +1292,8 @@ SinglePartitionJoin(JoinOrderNode *currentJoinNode, TableEntry *candidateTable,
 	}
 
 	/* evaluate re-partitioning the current table only if the rule didn't apply above */
-	if (nextJoinNode == NULL && candidatePartitionMethod != DISTRIBUTE_BY_HASH)
+	if (nextJoinNode == NULL && candidatePartitionMethod != DISTRIBUTE_BY_HASH &&
```

Comment:
We only do range-repartitioining for single-repartition joins and hash-repartitioning for dual-repartition joins. So when joining two hash-distributed tables, we always re-partition both even if one is partitioned on the join column. This is something we should consider fixing.